src/main/java/org/broadinstitute/hellbender/engine/ReadsPathDataSource.java

```diff
@@ -0,0 +1,520 @@
+
+import java.io.IOException;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Manages traversals and queries over sources of reads which are accessible via {@link Path}s
+ * (for now, SAM/BAM/CRAM files only).
+ *
+ * Two basic operations are available:
+ *
+ * -Iteration over all reads, optionally restricted to reads that overlap a set of intervals
+ * -Targeted queries by one interval at a time
+ */
+public final class ReadsPathDataSource implements ReadsDataSource {
+    private static final Logger logger = LogManager.getLogger(ReadsPathDataSource.class);
+
+    /**
+     * Mapping from SamReaders to iterators over the reads from each reader. Only one
+     * iterator can be open from a given reader at a time (this is a restriction
+     * in htsjdk). Iterator is set to null for a reader if no iteration is currently
+     * active on that reader.
+     */
+    private final Map<SamReader, CloseableIterator<SAMRecord>> readers;
+
+    /**
+     * Hang onto the input files so that we can print useful errors about them
+     */
+    private final Map<SamReader, Path> backingPaths;
+
+    /**
+     * Only reads that overlap these intervals (and unmapped reads, if {@link #traverseUnmapped} is set) will be returned
+     * during a full iteration. Null if iteration is unbounded.
+     *
+     * Individual queries are unaffected by these intervals -- only traversals initiated via {@link #iterator} are affected.
+     */
+    private List<SimpleInterval> intervalsForTraversal;
+
+    /**
+     * If true, restrict traversals to unmapped reads (and reads overlapping any {@link #intervalsForTraversal}, if set).
+     * False if iteration is unbounded or bounded only by our {@link #intervalsForTraversal}.
+     *
+     * Note that this setting covers only unmapped reads that have no position -- unmapped reads that are assigned the
+     * position of their mates will be returned by queries overlapping that position.
+     *
+     * Individual queries are unaffected by this setting  -- only traversals initiated via {@link #iterator} are affected.
+     */
+    private boolean traverseUnmapped;
+
+    /**
+     * Used to create a merged Sam header when we're dealing with multiple readers. Null if we only have a single reader.
+     */
+    private final SamFileHeaderMerger headerMerger;
+
+    /**
+     * Are indices available for all files?
+     */
+    private boolean indicesAvailable;
+
+    /**
+     * Initialize this data source with a single SAM/BAM file and validation stringency SILENT.
+     *
+     * @param samFile SAM/BAM file, not null.
+     */
+    public ReadsPathDataSource(final Path samFile ) {
```

Comment:
You have introduced your famous `(unbalanced parens )` in the `ReadsPathDataSource` constructors -- can you fix this?

Reply:
fixed